### PR TITLE
fix pyt 2.0 issues

### LINKED ISSUE #208: error: failed to legalize operation 'tt.load' that was explicitly marked illegal
**Pytorch anaylsis**
On the TOT triton-mlir (05/02/13) an issue was observed when running upstream pytorch UTs.
```
error: failed to legalize operation 'tt.load' that was explicitly marked illegal
Pass execution failedLLVM ERROR: Failed to translate TritonGPU to LLVM IR.
```

Here is the extracted triton codegen from one of the torch UT's
```
from ctypes import c_void_p, c_long
import torch
import math
import random
import os
import tempfile
from torch._inductor.hooks import run_intermediate_hooks
from torch._inductor.utils import maybe_profile

from torch import empty_strided, as_strided, device
from torch._inductor.codecache import AsyncCompile
from torch._inductor.select_algorithm import extern_kernels

aten = torch.ops.aten
assert_size_stride = torch._C._dynamo.guards.assert_size_stride
async_compile = AsyncCompile()

import triton
import triton.language as tl
from torch._inductor.triton_heuristics import grid, start_graph, end_graph
from torch._C import _cuda_getCurrentRawStream as get_cuda_stream


# kernel path: /tmp/torchinductor_root/gd/cgdifefxy25zaxpo46qpuwcca7dsvzzdgrny2roosqexbzqhmdmr.py
# Original ATen: aten.view

# aten.view => view_1
triton_poi_fused_view_0 = async_compile.triton('triton_', '''
import triton
import triton.language as tl
from torch._inductor.ir import ReductionHint
from torch._inductor.ir import TileHint
from torch._inductor.triton_heuristics import pointwise
from torch._inductor.utils import instance_descriptor
from torch._inductor import triton_helpers

@pointwise(size_hints=[1024], filename=__file__, meta={'signature': {0: '*i64', 1: '*fp32', 2: '*fp32', 3: 'i32'}, 'device': 0, 'constants': {}, 'mutated_arg_names': [], 'configs': [instance_descriptor(divisible_by_16=(0, 1, 2), equal_to_1=())]})
@triton.jit
def triton_(in_ptr0, in_ptr1, out_ptr0, xnumel, XBLOCK : tl.constexpr):
    xnumel = 600
    xoffset = tl.program_id(0) * XBLOCK
    xindex = xoffset + tl.arange(0, XBLOCK)[:]
    xmask = xindex < xnumel
    x2 = (xindex // 50) % 2
    x3 = (xindex // 100) % 3
    x4 = (xindex // 300)
    x5 = xindex % 50
    x0 = xindex % 5
    x1 = (xindex // 5) % 10
    x7 = xindex
    tmp0 = tl.load(in_ptr0 + ((2*x2) + (4*x3) + (4*((x5 + (50*x2)) // 100)) + (12*x4) + (12*((x5 + (50*x2) + (100*x3)) // 300)) + (x5 // 25)), xmask)
    tmp1 = tl.load(in_ptr1 + (x0 + (5*(x1 % 5)) + (25*tmp0)), xmask)
    tl.store(out_ptr0 + (tl.broadcast_to(x7, [XBLOCK])), tmp1, xmask)
''')

async_compile.wait(globals())
del async_compile

def call(args):
    arg0_1, arg1_1 = args
    args.clear()
    with torch.cuda._DeviceGuard(0):
        torch.cuda.set_device(0) # no-op to ensure context
        buf0 = empty_strided((2, 3, 2, 10, 5), (300, 100, 50, 5, 1), device='cuda', dtype=torch.float32)
        stream0 = get_cuda_stream(0)
        triton_poi_fused_view_0.run(arg1_1, arg0_1, buf0, 600, grid=grid(600), stream=stream0)
        run_intermediate_hooks('view_1', buf0)
        del arg0_1
        del arg1_1
        return (buf0, )


def benchmark_compiled_module(times=10, repeat=10):
    from torch._dynamo.testing import rand_strided
    from torch._inductor.utils import print_performance
    arg0_1 = rand_strided((12, 5, 5), (25, 5, 1), device='cuda:0', dtype=torch.float32)
    arg1_1 = rand_strided((24, ), (1, ), device='cuda:0', dtype=torch.int64)
    return print_performance(lambda: call([arg0_1, arg1_1]), times=times, repeat=repeat)


if __name__ == "__main__":
    from torch._inductor.utils import compiled_module_main
    compiled_module_main('None', benchmark_compiled_module)
```
After some experimentation it was found that setting TORCHINDUCTOR_COMPILE_THREADS=1 (effectively disabling multiprocessing) mitigated this issue and the torch UTs and workloads began passing.





**Triton reproducer**
The `test_cache` triton unit tests also use the multiprocessing module in the `test_compile_in_subproc` UT and can also replicate this error message.

```
HIP_VISIBLE_DEVICES=0 pytest test_cache.py  --verbose
platform linux -- Python 3.9.16, pytest-7.2.2, pluggy-1.0.0 -- /opt/conda/envs/py_3.9/bin/python
cachedir: .pytest_cache
hypothesis profile 'default' -> database=DirectoryBasedExampleDatabase('/var/lib/jenkins/triton/python/test/unit/runtime/.hypothesis/examples')
rootdir: /var/lib/jenkins/triton/python
plugins: hypothesis-5.35.1, xdoctest-1.1.0, xdist-3.2.1, shard-0.1.2, flakefinder-1.1.0, rerunfailures-11.1.2
collected 11 items                                                                                                                                                                                                                   
Running 11 items in this shard: test/unit/runtime/test_cache.py::test_nochange, test/unit/runtime/test_cache.py::test_toplevel_change, test/unit/runtime/test_cache.py::test_nested1_change, test/unit/runtime/test_cache.py::test_reuse, test/unit/runtime/test_cache.py::test_specialize[enable], test/unit/runtime/test_cache.py::test_specialize[disable], test/unit/runtime/test_cache.py::test_constexpr_not_callable, test/unit/runtime/test_cache.py::test_jit_warmup_cache, test/unit/runtime/test_cache.py::test_jit_debug, test/unit/runtime/test_cache.py::test_compile_in_subproc, test/unit/runtime/test_cache.py::test_memory_leak

test_cache.py::test_nochange PASSED                                                                                                                                                                                            [  9%]
test_cache.py::test_toplevel_change PASSED                                                                                                                                                                                     [ 18%]
test_cache.py::test_nested1_change PASSED                                                                                                                                                                                      [ 27%]
test_cache.py::test_reuse PASSED                                                                                                                                                                                               [ 36%]
test_cache.py::test_specialize[enable] PASSED                                                                                                                                                                                  [ 45%]
test_cache.py::test_specialize[disable] PASSED                                                                                                                                                                                 [ 54%]
test_cache.py::test_constexpr_not_callable PASSED                                                                                                                                                                              [ 63%]
test_cache.py::test_jit_warmup_cache PASSED                                                                                                                                                                                    [ 72%]
test_cache.py::test_jit_debug PASSED                                                                                                                                                                                           [ 81%]
test_cache.py::test_compile_in_subproc Fatal Python error: Aborted

Current thread 0x00007f9baa4fc4c0 (most recent call first):
  File "/var/lib/jenkins/triton/python/triton/compiler/compiler.py", line 103 in ttgir_to_llir
  File "/var/lib/jenkins/triton/python/triton/compiler/compiler.py", line 384 in <lambda>
  File "/var/lib/jenkins/triton/python/triton/compiler/compiler.py", line 465 in compile
  File "/opt/conda/envs/py_3.9/lib/python3.9/multiprocessing/process.py", line 108 in run
  File "/opt/conda/envs/py_3.9/lib/python3.9/multiprocessing/process.py", line 315 in _bootstrap
  File "/opt/conda/envs/py_3.9/lib/python3.9/multiprocessing/popen_fork.py", line 71 in _launch
  File "/opt/conda/envs/py_3.9/lib/python3.9/multiprocessing/popen_fork.py", line 19 in __init__
  File "/opt/conda/envs/py_3.9/lib/python3.9/multiprocessing/context.py", line 277 in _Popen
  File "/opt/conda/envs/py_3.9/lib/python3.9/multiprocessing/context.py", line 224 in _Popen
  File "/opt/conda/envs/py_3.9/lib/python3.9/multiprocessing/process.py", line 121 in start
  File "/var/lib/jenkins/triton/python/test/unit/runtime/test_cache.py", line 198 in test_compile_in_subproc
  File "/opt/conda/envs/py_3.9/lib/python3.9/site-packages/_pytest/python.py", line 195 in pytest_pyfunc_call
  File "/opt/conda/envs/py_3.9/lib/python3.9/site-packages/pluggy/_callers.py", line 39 in _multicall
  File "/opt/conda/envs/py_3.9/lib/python3.9/site-packages/pluggy/_manager.py", line 80 in _hookexec
  File "/opt/conda/envs/py_3.9/lib/python3.9/site-packages/pluggy/_hooks.py", line 265 in __call__
  File "/opt/conda/envs/py_3.9/lib/python3.9/site-packages/_pytest/python.py", line 1789 in runtest
  File "/opt/conda/envs/py_3.9/lib/python3.9/site-packages/_pytest/runner.py", line 167 in pytest_runtest_call
  File "/opt/conda/envs/py_3.9/lib/python3.9/site-packages/pluggy/_callers.py", line 39 in _multicall
  File "/opt/conda/envs/py_3.9/lib/python3.9/site-packages/pluggy/_manager.py", line 80 in _hookexec
  File "/opt/conda/envs/py_3.9/lib/python3.9/site-packages/pluggy/_hooks.py", line 265 in __call__
  File "/opt/conda/envs/py_3.9/lib/python3.9/site-packages/_pytest/runner.py", line 260 in <lambda>
  File "/opt/conda/envs/py_3.9/lib/python3.9/site-packages/_pytest/runner.py", line 339 in from_call
  File "/opt/conda/envs/py_3.9/lib/python3.9/site-packages/_pytest/runner.py", line 259 in call_runtest_hook
  File "/opt/conda/envs/py_3.9/lib/python3.9/site-packages/_pytest/runner.py", line 220 in call_and_report
  File "/opt/conda/envs/py_3.9/lib/python3.9/site-packages/_pytest/runner.py", line 131 in runtestprotocol
  File "/opt/conda/envs/py_3.9/lib/python3.9/site-packages/_pytest/runner.py", line 112 in pytest_runtest_protocol
  File "/opt/conda/envs/py_3.9/lib/python3.9/site-packages/pluggy/_callers.py", line 39 in _multicall
  File "/opt/conda/envs/py_3.9/lib/python3.9/site-packages/pluggy/_manager.py", line 80 in _hookexec
  File "/opt/conda/envs/py_3.9/lib/python3.9/site-packages/pluggy/_hooks.py", line 265 in __call__
  File "/opt/conda/envs/py_3.9/lib/python3.9/site-packages/_pytest/main.py", line 349 in pytest_runtestloop
  File "/opt/conda/envs/py_3.9/lib/python3.9/site-packages/pluggy/_callers.py", line 39 in _multicall
  File "/opt/conda/envs/py_3.9/lib/python3.9/site-packages/pluggy/_manager.py", line 80 in _hookexec
  File "/opt/conda/envs/py_3.9/lib/python3.9/site-packages/pluggy/_hooks.py", line 265 in __call__
  File "/opt/conda/envs/py_3.9/lib/python3.9/site-packages/_pytest/main.py", line 324 in _main
  File "/opt/conda/envs/py_3.9/lib/python3.9/site-packages/_pytest/main.py", line 270 in wrap_session
  File "/opt/conda/envs/py_3.9/lib/python3.9/site-packages/_pytest/main.py", line 317 in pytest_cmdline_main
  File "/opt/conda/envs/py_3.9/lib/python3.9/site-packages/pluggy/_callers.py", line 39 in _multicall
  File "/opt/conda/envs/py_3.9/lib/python3.9/site-packages/pluggy/_manager.py", line 80 in _hookexec
  File "/opt/conda/envs/py_3.9/lib/python3.9/site-packages/pluggy/_hooks.py", line 265 in __call__
  File "/opt/conda/envs/py_3.9/lib/python3.9/site-packages/_pytest/config/__init__.py", line 167 in main
  File "/opt/conda/envs/py_3.9/lib/python3.9/site-packages/_pytest/config/__init__.py", line 190 in console_main
  File "/opt/conda/envs/py_3.9/bin/pytest", line 8 in <module>
FAILED                                                                                                                                                                                  [ 90%]
test_cache.py::test_memory_leak PASSED                                                                                                                                                                                         [100%]

============================================================================================================== FAILURES ==============================================================================================================
______________________________________________________________________________________________________ test_compile_in_subproc _______________________________________________________________________________________________________

    def test_compile_in_subproc() -> None:
        @triton.jit
        def kernel_sub(a, b, o, N: tl.constexpr):
            idx = tl.arange(0, N)
            tl.store(o + idx,
                     tl.load(a + idx) - tl.load(b + idx) * 777)
    
        major, minor = torch.cuda.get_device_capability(0)
        cc = major * 10 + minor
        config = namedtuple("instance_descriptor", [
            "divisible_by_16", "equal_to_1"])(
            tuple(range(4)),
            ())
    
        proc = multiprocessing.Process(
            target=triton.compile,
            kwargs=dict(
                fn=kernel_sub,
                signature={0: "*fp32", 1: "*fp32", 2: "*fp32"},
                device=0,
                constants={3: 32},
                configs=[config],
                warm_cache_only=True,
                cc=cc,
            ))
        proc.start()
        proc.join()
>       assert proc.exitcode == 0
E       AssertionError: assert -6 == 0
E        +  where -6 = <Process name='Process-1' pid=2017813 parent=2017544 stopped exitcode=-SIGABRT>.exitcode

test_cache.py:200: AssertionError
Captured stderr call 
error: failed to legalize operation 'tt.load' that was explicitly marked illegal
Pass execution failedLLVM ERROR: Failed to translate TritonGPU to LLVM IR.
FAILED test_cache.py::test_compile_in_subproc - AssertionError: assert -6 == 0
```

cc: 
@dllehr-amd @jayfurmanek @micmelesse @zstreet87

### PR DESCRIPTION
**Pull request by micmelesse**
fixes #208

### CODE REVIEW

**Comment by micmelesse** [comment]
@jataylo I have cleaned up the pr. Can you test that it still fixes the issue.

**Review by jataylo — APPROVED** [review]

**Review by jayfurmanek — APPROVED** [review]

File: scripts/amd/docker_run.sh [hunk 1]
```diff
@@ -12,11 +12,9 @@ VOLUMES="-v $HOME/dockerx:/dockerx -v /data:/data"
 # WORK_DIR='/root/$(basename $(pwd))'
 WORK_DIR="/dockerx/$(basename $(pwd))"
 
-# IMAGE_NAME=nvcr.io/nvidia/pytorch:21.08-py3
-# IMAGE_NAME=rocm/pytorch:latest 
-IMAGE_NAME=rocm/pytorch:rocm5.4_ubuntu20.04_py3.8_pytorch_1.12.1
-# IMAGE_NAME=rocm/pytorch:rocm4.3.1_ubuntu18.04_py3.6_pytorch_1.10.0
-# IMAGE_NAME=triton_rocm_20-52 # build this docker before running
+IMAGE_NAME=rocm/pytorch-nightly:latest
```
Comment (dllehr-amd) [inline scripts/amd/docker_run.sh:15]: For these runs...Should we try to pin a pytorch level so you don't get false positives if something else happens with Pytorch unrelated to this particular problem?

Comment (micmelesse) [reply scripts/amd/docker_run.sh:15]: this is a helper script. It is not used by the ci. The CI uses pulls the torch wheel. See [here](https://github.com/ROCmSoftwarePlatform/triton/blob/aabc2511f6553c2f26eb9c2aec7cf25f6421baa6/.github/workflows/amd-offline-tests.yml#L52)

File: python/triton/tools/aot.py [hunk 2]
```diff
@@ -83,24 +83,31 @@
...
 
-    if not args.sm:
-        raise argparse.ArgumentError(None, "Must specify --sm for PTX compilation")
+    # set arch depending on platform
+    if args.gfx:
+        arch = args.gfx
+    elif args.sm:
+        arch = args.sm
+    else:
+        raise argparse.ArgumentError(None, "Must specify --sm or --gfx for ttgir compilation")
```
Comment (binarman) [inline python/triton/tools/aot.py:92]: Consider refactor code above too: https://github.com/ROCmSoftwarePlatform/triton/pull/210/files#diff-3df9553c27ea5bf270c80c9193bb384cbe2561a669ac04a65918124bd83aff6fR49-R84

It masks AMD GPU cases, so this changes is unused.

Comment (micmelesse) [reply python/triton/tools/aot.py:92]: I am not sure I understand. I added this as a fix for this[ lit test](https://github.com/ROCmSoftwarePlatform/triton/pull/210/files#diff-8d11e23ee6ecc001d8181f4b28b20b569489e31779837e068b7709b8c5a7d85dR1). The issue is related to the path taken by when the target is llvmir. The AMDGCN path exists [before ](https://github.com/ROCmSoftwarePlatform/triton/blob/275fead8e3dee1d9e0b7201cfcccc5de24685466/python/triton/tools/aot.py#L84 )this code is hit. Can you add more detail?

**Pull request merged**